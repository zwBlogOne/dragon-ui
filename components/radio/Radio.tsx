import React, { Component } from 'react';
import classnames from 'classnames';
import PropsType from './PropsType';

class Radio extends Component<PropsType, any> {

  static defaultProps = {
    defaultChecked: false,
    isDisabled: false,
    onChange: () => {},
  };

  static Group;

  constructor(props) {
    super(props);
    this.state = {
      checked: props.checked || props.defaultChecked,
    };
  }

  componentWillReceiveProps(nextProps) {
    if ('checked' in nextProps) {
      this.setState({
        checked: !!nextProps.checked,
      });
    }
  }

  _onClick(e) {
    this.setState({
      checked: true,
    });
    this.props.onChange(e);
  }

  render() {
    const { props } = this;
    const {
      value, isDisabled, className, children,
    } = props;
    const disabled = 'disabled' in props || isDisabled;

    const cls = classnames({
      'ui-radio': true,
      checked: this.state.checked,
      disabled,
      [className!]: !!className,
    });

    return (
      <label>
        <span className={cls}>
          <input
            className="ui-radio-input"
            type="radio"
            value={value}
            checked={this.state.checked}
            disabled={disabled}
            onChange={e => this._onClick(e)}
          />
          <span className="ui-radio-inner" />
        </span>
        {children}
      </label>
    );
  }
}

export default Radio;
