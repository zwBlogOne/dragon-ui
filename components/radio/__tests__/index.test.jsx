import React from 'react';
import { render, mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import Radio from '../index';

describe('Radio', () => {
  it('renders normal Radio correctly', () => {
    const wrapper = render(
      <div>
        <Radio
          value="a"
        >
          选择
        </Radio>
      </div>
    );

    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('renders disabled Radio correctly', () => {
    const wrapper = render(
      <div>
        <Radio
          disabled
          value="a"
        >
          选择
        </Radio>
      </div>
    );

    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('renders checked Radio correctly', () => {
    const wrapper = render(
      <div>
        <Radio
          checked
        >
          选择
        </Radio>
      </div>
    );

    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('renders Radio Group correctly', () => {
    const wrapper = render(
      <Radio.Group
        value="a"
      >
        <Radio value="a">A</Radio>
        <Radio value="b">B</Radio>
        <Radio value="c">C</Radio>
        <Radio value="d">D</Radio>
      </Radio.Group>
    );

    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('behaves correctly when changing radio status', () => {
    const onChange = jest.fn();
    const wrapper = mount(
      <Radio
        value="a"
        onChange={onChange}
      >
        选择
      </Radio>
    );

    wrapper.find('.ui-radio-input').simulate('change');
    expect(onChange).toBeCalled();
  });

  it('behaves correctly when choosing from Radio Group', () => {
    const onChange = jest.fn();
    const wrapper = mount(
      <Radio.Group
        value="a"
        onChange={onChange}
      >
        <Radio value="a">A</Radio>
        <Radio value="b">B</Radio>
        <Radio value="c">C</Radio>
        <Radio value="d">D</Radio>
      </Radio.Group>
    );

    wrapper.find('.ui-radio-input').first().simulate('change');
    expect(onChange).toBeCalled();
  });
});

