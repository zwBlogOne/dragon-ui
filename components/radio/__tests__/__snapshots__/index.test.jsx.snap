// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Radio renders Radio Group correctly 1`] = `
<div
  class="ui-radio-group"
>
  <label>
    <span
      class="ui-radio checked"
    >
      <input
        checked=""
        class="ui-radio-input"
        type="radio"
        value="a"
      />
      <span
        class="ui-radio-inner"
      />
    </span>
    A
  </label>
  <label>
    <span
      class="ui-radio"
    >
      <input
        class="ui-radio-input"
        type="radio"
        value="b"
      />
      <span
        class="ui-radio-inner"
      />
    </span>
    B
  </label>
  <label>
    <span
      class="ui-radio"
    >
      <input
        class="ui-radio-input"
        type="radio"
        value="c"
      />
      <span
        class="ui-radio-inner"
      />
    </span>
    C
  </label>
  <label>
    <span
      class="ui-radio"
    >
      <input
        class="ui-radio-input"
        type="radio"
        value="d"
      />
      <span
        class="ui-radio-inner"
      />
    </span>
    D
  </label>
</div>
`;

exports[`Radio renders checked Radio correctly 1`] = `
<div>
  <label>
    <span
      class="ui-radio checked"
    >
      <input
        checked=""
        class="ui-radio-input"
        type="radio"
      />
      <span
        class="ui-radio-inner"
      />
    </span>
    选择
  </label>
</div>
`;

exports[`Radio renders disabled Radio correctly 1`] = `
<div>
  <label>
    <span
      class="ui-radio disabled"
    >
      <input
        class="ui-radio-input"
        disabled=""
        type="radio"
        value="a"
      />
      <span
        class="ui-radio-inner"
      />
    </span>
    选择
  </label>
</div>
`;

exports[`Radio renders normal Radio correctly 1`] = `
<div>
  <label>
    <span
      class="ui-radio"
    >
      <input
        class="ui-radio-input"
        type="radio"
        value="a"
      />
      <span
        class="ui-radio-inner"
      />
    </span>
    选择
  </label>
</div>
`;
